Runtime/Scripts/BlendShapesController.cs

```diff
@@ -175,87 +203,100 @@ public void UpdateBlendShapeIndices(IStreamSettings settings)
                 }
 
                 m_Indices.Add(meshRenderer, indices);
-            }
+            });
         }
 
-        /// <summary>
-        /// Calculate what the interpolated blend shape values should be before they are set to the rig.
-        /// </summary>
-        /// <param name="force">Force smoothing.</param>
-        public void CalculateInterpolatedBlendShapes(bool force = false)
+        void PostProcessValues(float deltaTime)
         {
-            var streamSettings = streamReader.streamSource.streamSettings;
-            for (var i = 0; i < streamSettings.BlendShapeCount; i++)
+            for (var i = 0; i < m_BlendShapeValues.Count; i++)
             {
-                var blendShape = m_BlendShapes[i];
-                var blendShapeTarget = streamReader.blendShapesBuffer[i];
-                var useOverride = UseOverride(i);
-                var blendShapeOverride = m_Overrides[i];
-                var threshold = useOverride ? blendShapeOverride.blendShapeThreshold : m_BlendShapeThreshold;
-                var offset = useOverride ? blendShapeOverride.blendShapeOffset : 0f;
-                var smoothing = useOverride ? blendShapeOverride.blendShapeSmoothing : m_BlendShapeSmoothing;
-
-                if (force || !streamReader.faceTrackingLost)
+                var targetValue = m_BlendShapeValues[i];
+                var hasOverride = HasOverride(i);
+                var blendShapeOverride = m_Overrides[i]; //TODO: this might crash
+                var threshold = hasOverride ? blendShapeOverride.blendShapeThreshold : m_Threshold;
+                var offset = hasOverride ? blendShapeOverride.blendShapeOffset : 0f;
+                var smoothing = hasOverride ? blendShapeOverride.blendShapeSmoothing : m_Smoothing;
+                var scale = hasOverride ? blendShapeOverride.blendShapeCoefficient : m_Multiplier;
+                var maxValue = hasOverride ? blendShapeOverride.blendShapeMax : m_Maximum;
+                var currentValue = m_SmoothedBlendShapes[i].x;
+                var currentSpeed = m_SmoothedBlendShapes[i].y;
+
+                if (isTrackingActive)
                 {
-                    if (Mathf.Abs(blendShapeTarget - blendShape) > threshold)
-                        m_BlendShapes[i] = Mathf.Lerp(blendShapeTarget, blendShape, smoothing);
+                    if (Mathf.Abs(targetValue - currentValue) > threshold)
+                        currentValue = Mathf.SmoothDamp(currentValue, targetValue, ref currentSpeed, smoothing, Mathf.Infinity, deltaTime);
```

Comment:
Why SmoothDamp? Can you provide any mathematical justification? I'm worried about damping unnecessarily? How is this done in something like Motion Builder or something?

Reply:
I had to replace the current Lerp function with something that is time-dependent. If we bake the results of the Lerp implementation we will observe that the smoothing will be dependent on the sample rate of the clip. We could use smoothing * smoothingVelocity * deltaTime as t, but then we introduce a new parameter and still the results won't be smooth.
SmoothDamp implements a critically damped movement (no oscillations), it is time-dependent and it is smooth in both velocity and position with no extra parameters.
(Game Programming Gems 4 Chapter 1.10)

Reply:
Reverted to Lerp